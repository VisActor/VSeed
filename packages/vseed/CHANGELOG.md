# @visactor/vseed

## 0.0.28

### Patch Changes

- feat: remove d3 color

## 0.0.27

### Patch Changes

- fix: zod types err

## 0.0.26

### Patch Changes

- fix: same alias error in parallel chart

## 0.0.25

### Patch Changes

- feat: add heatmap
- feat: optimize the zero dimensions style in line\area chart
- fix: stroke error in pie and donut
- feat: optimize the linear color in funnel

## 0.0.24

### Patch Changes

- feat: add radar chart

## 0.0.23

### Patch Changes

- feat: add area range and update spec api

## 0.0.22

### Patch Changes

- feat: add axis sort and legend sort

## 0.0.21

### Patch Changes

- feat: add table and pivot table

## 0.0.20

### Patch Changes

- feat: add table and pivot table

## 0.0.19

### Patch Changes

- breaking-change:
  - theme, move all base config(color, legend, label, tooltip, backgroundColor) to config
  - format, remove round, decimalPlaces

  feat:
  - num format add fractionDigits, significantDigits, roundingPriority,roundingMode
  - linear log and log base

## 0.0.18

### Patch Changes

- fix: data reshape err

## 0.0.17

### Patch Changes

- feat: add scatter, donut, funnel
- feat: optimize line sort
- feat: optimize color mapping

## 0.0.16

### Patch Changes

- feat: add rose roseParallel donut

## 0.0.15

### Patch Changes

- feat: add zVSeed

## 0.0.14

### Patch Changes

- feat: allow same alias
- feat: legend label format
- fix: alias missing error
- docs: optimize guide

## 0.0.13

### Patch Changes

- feat: auto format
- feat: i18n and locale in vseed

## 0.0.12

### Patch Changes

- feat: add measure format
- feat: optimize light and dark theme
- feat: add line style
- feat: add area style
- feat: add point style
- feat: optimize line and bar hover style
- docs: demos for new features

## 0.0.11

### Patch Changes

- feat: add annotation line
- feat: add annotation area
- docs: optimize the guide and demos

## 0.0.10

### Patch Changes

- feat: add annotation point

## 0.0.9

### Patch Changes

- feat: add bar style
- feat: optimize legend
- feat: optimize theme

## 0.0.8

### Patch Changes

- feat: add x and y axis

## 0.0.7

### Patch Changes

- feat: add pie chart

## 0.0.6

### Patch Changes

- feat: combination chart

## 0.0.5

### Patch Changes

- feat: add theme and base config

## 0.0.4

### Patch Changes

- feat: theme and base config

## 0.0.3

### Patch Changes

- release 0.0.3

## 0.0.2

### Patch Changes

- feat: add column and bar chart

## 0.0.1

### Patch Changes

- test release
