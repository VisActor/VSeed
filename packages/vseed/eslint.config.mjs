import js from '@eslint/js'
import globals from 'globals'
import ts from 'typescript-eslint'
import tseslint from 'typescript-eslint'

export default [
  { languageOptions: { globals: globals.browser } },
  js.configs.recommended,
  ...ts.configs.recommended,
  { ignores: ['dist/'] },

  ...tseslint.configs.recommendedTypeChecked,
  {
    files: ['src/**/*.ts', 'tests/**/*.ts'],
    languageOptions: {
      parserOptions: {
        projectService: true,
        tsconfigRootDir: import.meta.dirname,
      },
    },
    rules: {
      '@typescript-eslint/consistent-type-imports': 'error',
      '@typescript-eslint/consistent-type-exports': 'error',
      '@typescript-eslint/no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
    },
  },
]
