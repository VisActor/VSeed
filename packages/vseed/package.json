{
  "name": "@visactor/vseed",
  "version": "0.1.0",
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.js",
      "require": "./dist/index.cjs"
    }
  },
  "main": "./dist/index.cjs",
  "types": "./dist/index.d.ts",
  "unpkg": "./dist/umd/index.js",
  "files": [
    "dist"
  ],
  "scripts": {
    "dev": "rslib build --watch",
    "build": "rslib build",
    "dev:rsdoctor": "RSDOCTOR=true rsbuild dev",
    "build:rsdoctor": "RSDOCTOR=true rsbuild build",
    "format": "prettier --write .",
    "lint": "eslint .",
    "test": "vitest run",
    "rstest": "rstest",
    "test:coverage": "vitest run --coverage"
  },
  "dependencies": {
    "remeda": "^2.28.0",
    "zod": "4.0.14"
  },
  "sideEffects": false,
  "devDependencies": {
    "@eslint/js": "9.30.0",
    "@rsdoctor/rspack-plugin": "1.1.10",
    "@rslib/core": "0.12.0",
    "@rstest/core": "^0.3.2",
    "@types/d3-color": "^3.1.0",
    "@types/node": "^22.16.5",
    "@typescript-eslint/eslint-plugin": "8.38.0",
    "@visactor/vchart": "1.13.19",
    "@visactor/vtable": "1.19.8",
    "@vitest/coverage-istanbul": "3.2.4",
    "@vitest/coverage-v8": "3.2.4",
    "eslint": "9.30.0",
    "globals": "16.2.0",
    "jsdom": "^26.1.0",
    "rstest-canvas-mock": "^0.0.1",
    "typescript": "5.8.3",
    "typescript-eslint": "8.35.1",
    "vitest": "3.2.4",
    "vitest-canvas-mock": "^0.3.3"
  }
}
