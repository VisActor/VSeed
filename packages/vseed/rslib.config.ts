import { defineConfig } from '@rslib/core'

export default defineConfig({
  lib: [
    {
      format: 'esm',
      syntax: ['node 18'],
      dts: true,
    },
    {
      format: 'cjs',
      syntax: ['node 18'],
    },
    {
      format: 'umd',
      umdName: 'VSeed',
      performance:{
        chunkSplit:{
          strategy: 'all-in-one',
        }
      },
      output: {
        target: 'web',
        distPath: {
          root: './dist/umd',
        },
      },
    },
  ],
  output: {
    sourceMap: true,
  },
})
