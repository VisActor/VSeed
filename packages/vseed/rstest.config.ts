import { defineConfig } from '@rstest/core'

export default defineConfig({
  setupFiles: ['./rstest.setup.ts'],
  globals: true,
  testEnvironment: 'jsdom',
})
