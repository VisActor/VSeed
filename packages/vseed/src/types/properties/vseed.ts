import z from 'zod'
import type { ChartType } from './chartType'
import { zChartType } from './chartType'
import type { Dataset } from './dataset'
import { zDataset } from './dataset'
import type { Dimensions } from './dimensions'
import { zDimensions } from './dimensions'
import type { Measures } from './measures'
import { zMeasures } from './measures'

export const zVSeed = z.object({
  chartType: zChartType,
  dataset: zDataset,
  dimensions: zDimensions,
  measures: zMeasures,
})

export type VSeed = {
  chartType: ChartType
  dataset: Dataset
  dimensions: Dimensions
  measures: Measures
}

export const VSeedJSONSchema = z.toJSONSchema(zVSeed)
