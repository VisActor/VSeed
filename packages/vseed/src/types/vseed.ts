import { z } from 'zod'

import { zBar, type Bar } from './chartType/bar'
import { zBarParallel, type BarParallel } from './chartType/barParallel'
import { zColumn, type Column } from './chartType/column'
import { zColumnParallel, type ColumnParallel } from './chartType/columnParallel'
import { zColumnPercent, type ColumnPercent } from './chartType/columnPercent'
import { zLine, type Line } from './chartType/line'
import type { Table } from './chartType/table'
import type { Rose } from './chartType/rose'
import { zBarPercent, type BarPercent } from './chartType/barPercent'
import type { PivotTable } from './chartType/pivotTable'
import { zPie, type Pie } from './chartType/pie'
import type { Donut } from './chartType/donut'
import type { DualAxis } from './chartType/dualAxis'
import { zArea, type Area } from './chartType/area'
import { zAreaPercent, type AreaPercent } from './chartType/areaPercent'

export type VSeed =
  | Table
  | PivotTable
  | Line
  | Column
  | ColumnParallel
  | ColumnPercent
  | Bar
  | BarParallel
  | BarPercent
  | Area
  | AreaPercent
  | DualAxis
  | Rose
  | Pie
  | Donut

export const zVSeed = z.discriminatedUnion('chartType', [
  zLine,
  zColumn,
  zColumnParallel,
  zColumnPercent,
  zBar,
  zBarParallel,
  zBarPercent,
  zArea,
  zAreaPercent,
  zPie,
  // zRose,
  // zDualAxis,
  // zDonut,
])
