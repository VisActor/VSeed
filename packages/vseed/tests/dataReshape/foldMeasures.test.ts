import type { Dataset, Measure } from '@visactor/vseed'
import { foldMeasures } from '@visactor/vseed'

describe('foldMeasures', () => {
  test('0 dimensions, 0 measures', () => {
    const dataset: Dataset = []
    const measures: Measure[] = []
    const result = foldMeasures(dataset, measures)
    const newDataset = result.dataset
    expect(newDataset.length).toBe(dataset.length * measures.length)
    expect(result).toMatchSnapshot()
  })

  test('0 dimensions, 1 measures', () => {
    const dataset: Dataset = [{ profit: 10 }]
    const measures: Measure[] = [{ id: 'profit', alias: 'Profit' }]
    const result = foldMeasures(dataset, measures)
    const newDataset = result.dataset
    expect(newDataset.length).toBe(dataset.length * measures.length)
    expect(result).toMatchSnapshot()
  })

  test('0 dimensions, 2 measures', () => {
    const dataset: Dataset = [{ profit: 10, sales: 100 }]
    const measures: Measure[] = [
      { id: 'profit', alias: 'Profit' },
      { id: 'sales', alias: 'Sales' },
    ]
    const result = foldMeasures(dataset, measures)
    const newDataset = result.dataset
    expect(newDataset.length).toBe(dataset.length * measures.length)
    expect(result).toMatchSnapshot()
  })

  test('0 dimensions, 5 measures', () => {
    const dataset: Dataset = [{ profit: 10, sales: 100, number: 1, price: 100, cost: 33 }]
    const measures: Measure[] = [
      { id: 'profit', alias: 'Profit' },
      { id: 'sales', alias: 'Sales' },
      { id: 'number', alias: 'Number' },
      { id: 'price', alias: 'Price' },
      { id: 'cost', alias: 'Cost' },
    ]
    const result = foldMeasures(dataset, measures)
    const newDataset = result.dataset
    expect(newDataset.length).toBe(dataset.length * measures.length)
    expect(result).toMatchSnapshot()
  })

  test('1 dimensions, 1 measures', () => {
    const dataset: Dataset = [
      { date: '2019', sales: 100 },
      { date: '2020', sales: 3200 },
      { date: '2021', sales: 300 },
      { date: '2022', sales: 2400 },
      { date: '2023', sales: 500 },
      { date: '2019', sales: 100 },
      { date: '2020', sales: 3200 },
      { date: '2021', sales: 300 },
      { date: '2022', sales: 2400 },
      { date: '2023', sales: 500 },
    ]
    const measures: Measure[] = [{ id: 'profit', alias: 'Profit' }]
    const result = foldMeasures(dataset, measures)
    const newDataset = result.dataset
    expect(newDataset.length).toBe(dataset.length * measures.length)
    expect(result).toMatchSnapshot()
  })

  test('2 dimensions, 1 measures', () => {
    const dataset: Dataset = [
      { date: '2019', type: 'A', profit: 10, sales: 100 },
      { date: '2020', type: 'A', profit: 30, sales: 3200 },
      { date: '2021', type: 'A', profit: 30, sales: 300 },
      { date: '2022', type: 'A', profit: 50, sales: 2400 },
      { date: '2023', type: 'A', profit: 40, sales: 500 },
      { date: '2019', type: 'B', profit: 10, sales: 100 },
      { date: '2020', type: 'B', profit: 30, sales: 3200 },
      { date: '2021', type: 'B', profit: 30, sales: 300 },
      { date: '2022', type: 'B', profit: 50, sales: 2400 },
      { date: '2023', type: 'B', profit: 40, sales: 500 },
    ]
    const measures: Measure[] = [{ id: 'profit', alias: 'Profit' }]
    const result = foldMeasures(dataset, measures)
    const newDataset = result.dataset
    expect(newDataset.length).toBe(dataset.length * measures.length)
    expect(result).toMatchSnapshot()
  })

  test('2 dimensions, 2 measures', () => {
    const dataset: Dataset = [
      { date: '2019', type: 'A', profit: 10, sales: 100 },
      { date: '2020', type: 'A', profit: 30, sales: 3200 },
      { date: '2021', type: 'A', profit: 30, sales: 300 },
      { date: '2022', type: 'A', profit: 50, sales: 2400 },
      { date: '2023', type: 'A', profit: 40, sales: 500 },
      { date: '2019', type: 'B', profit: 10, sales: 100 },
      { date: '2020', type: 'B', profit: 30, sales: 3200 },
      { date: '2021', type: 'B', profit: 30, sales: 300 },
      { date: '2022', type: 'B', profit: 50, sales: 2400 },
      { date: '2023', type: 'B', profit: 40, sales: 500 },
    ]
    const measures: Measure[] = [
      { id: 'profit', alias: 'Profit' },
      { id: 'sales', alias: 'Sales' },
    ]
    const result = foldMeasures(dataset, measures)
    const newDataset = result.dataset
    expect(newDataset.length).toBe(dataset.length * measures.length)
    expect(result).toMatchSnapshot()
  })
})
