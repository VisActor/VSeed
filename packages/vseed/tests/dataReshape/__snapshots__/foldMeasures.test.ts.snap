// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`foldMeasures > 0 dimensions, 0 measures 1`] = `
{
  "dataset": [],
  "foldInfo": {
    "foldMap": {},
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;

exports[`foldMeasures > 0 dimensions, 1 measures 1`] = `
{
  "dataset": [
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 10,
      "profit": 10,
    },
  ],
  "foldInfo": {
    "foldMap": {
      "profit": "Profit",
    },
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;

exports[`foldMeasures > 0 dimensions, 2 measures 1`] = `
{
  "dataset": [
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "profit": 10,
      "sales": 100,
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "profit": 10,
      "sales": 100,
    },
  ],
  "foldInfo": {
    "foldMap": {
      "profit": "Profit",
      "sales": "Sales",
    },
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;

exports[`foldMeasures > 0 dimensions, 5 measures 1`] = `
{
  "dataset": [
    {
      "__MeaId__": "cost",
      "__MeaName__": "Cost",
      "__MeaValue__": 33,
      "cost": 33,
      "number": 1,
      "price": 100,
      "profit": 10,
      "sales": 100,
    },
    {
      "__MeaId__": "cost",
      "__MeaName__": "Cost",
      "__MeaValue__": 33,
      "cost": 33,
      "number": 1,
      "price": 100,
      "profit": 10,
      "sales": 100,
    },
    {
      "__MeaId__": "cost",
      "__MeaName__": "Cost",
      "__MeaValue__": 33,
      "cost": 33,
      "number": 1,
      "price": 100,
      "profit": 10,
      "sales": 100,
    },
    {
      "__MeaId__": "cost",
      "__MeaName__": "Cost",
      "__MeaValue__": 33,
      "cost": 33,
      "number": 1,
      "price": 100,
      "profit": 10,
      "sales": 100,
    },
    {
      "__MeaId__": "cost",
      "__MeaName__": "Cost",
      "__MeaValue__": 33,
      "cost": 33,
      "number": 1,
      "price": 100,
      "profit": 10,
      "sales": 100,
    },
  ],
  "foldInfo": {
    "foldMap": {
      "cost": "Cost",
      "number": "Number",
      "price": "Price",
      "profit": "Profit",
      "sales": "Sales",
    },
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;

exports[`foldMeasures > 1 dimensions, 1 measures 1`] = `
{
  "dataset": [
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2019",
      "sales": 100,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2020",
      "sales": 3200,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2021",
      "sales": 300,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2022",
      "sales": 2400,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2023",
      "sales": 500,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2019",
      "sales": 100,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2020",
      "sales": 3200,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2021",
      "sales": 300,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2022",
      "sales": 2400,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2023",
      "sales": 500,
    },
  ],
  "foldInfo": {
    "foldMap": {
      "profit": "Profit",
    },
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;

exports[`foldMeasures > 2 dimensions, 1 measures 1`] = `
{
  "dataset": [
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 10,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "A",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 30,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "A",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 30,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "A",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 50,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "A",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 40,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "A",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 10,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "B",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 30,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "B",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 30,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "B",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 50,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "B",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 40,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "B",
    },
  ],
  "foldInfo": {
    "foldMap": {
      "profit": "Profit",
    },
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;

exports[`foldMeasures > 2 dimensions, 2 measures 1`] = `
{
  "dataset": [
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 3200,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 3200,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 300,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 300,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 2400,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 2400,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 500,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 500,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 3200,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 3200,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 300,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 300,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 2400,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 2400,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 500,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 500,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "B",
    },
  ],
  "foldInfo": {
    "foldMap": {
      "profit": "Profit",
      "sales": "Sales",
    },
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;

exports[`foldMeasures > foldMeasures 0 dimensions 2 measures 1`] = `
{
  "dataset": [
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "profit": 10,
      "sales": 100,
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "profit": 10,
      "sales": 100,
    },
  ],
  "foldInfo": {
    "foldMap": {
      "profit": "Profit",
      "sales": "Sales",
    },
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;

exports[`foldMeasures > foldMeasures 1 dimensions 1 measures 1`] = `
{
  "dataset": [
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2019",
      "sales": 100,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2020",
      "sales": 3200,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2021",
      "sales": 300,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2022",
      "sales": 2400,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2023",
      "sales": 500,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2019",
      "sales": 100,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2020",
      "sales": 3200,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2021",
      "sales": 300,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2022",
      "sales": 2400,
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": undefined,
      "date": "2023",
      "sales": 500,
    },
  ],
  "foldInfo": {
    "foldMap": {
      "profit": "Profit",
    },
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;

exports[`foldMeasures > foldMeasures 2 dimensions 1 measures 1`] = `
{
  "dataset": [
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 10,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "A",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 30,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "A",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 30,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "A",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 50,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "A",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 40,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "A",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 10,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "B",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 30,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "B",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 30,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "B",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 50,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "B",
    },
    {
      "__MeaId__": "profit",
      "__MeaName__": "Profit",
      "__MeaValue__": 40,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "B",
    },
  ],
  "foldInfo": {
    "foldMap": {
      "profit": "Profit",
    },
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;

exports[`foldMeasures > foldMeasures 2 dimensions 2 measures 1`] = `
{
  "dataset": [
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 3200,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 3200,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 300,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 300,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 2400,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 2400,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 500,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 500,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "A",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 100,
      "date": "2019",
      "profit": 10,
      "sales": 100,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 3200,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 3200,
      "date": "2020",
      "profit": 30,
      "sales": 3200,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 300,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 300,
      "date": "2021",
      "profit": 30,
      "sales": 300,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 2400,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 2400,
      "date": "2022",
      "profit": 50,
      "sales": 2400,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 500,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "B",
    },
    {
      "__MeaId__": "sales",
      "__MeaName__": "Sales",
      "__MeaValue__": 500,
      "date": "2023",
      "profit": 40,
      "sales": 500,
      "type": "B",
    },
  ],
  "foldInfo": {
    "foldMap": {
      "profit": "Profit",
      "sales": "Sales",
    },
    "measureId": "__MeaId__",
    "measureName": "__MeaName__",
    "measureValue": "__MeaValue__",
  },
}
`;
